libsolidity/analysis/TypeChecker.cpp

```diff
@@ -1478,13 +1478,13 @@ bool TypeChecker::visit(UnaryOperation const& _operation)
 	TypePointer t = type(_operation.subExpression())->unaryOperatorResult(op);
 	if (!t)
 	{
-		m_errorReporter.typeError(
-			_operation.location(),
-			"Unary operator " +
-			string(TokenTraits::toString(op)) +
-			" cannot be applied to type " +
-			subExprType->toString()
-		);
+		string description = "Unary operator " + string(TokenTraits::toString(op)) + " cannot be applied to type " + subExprType->toString();
+		if (modifying)
```

Comment:
I think we should be much more liberal with applying fatal errors than we have been before.

Reply:
I could use `fatalTypeError` only, but then I would need to convert https://github.com/ethereum/solidity/blob/develop/test/libsolidity/syntaxTests/types/bool_ops.sol into several smaller tests. I would be fine with that too, no strong opinion here.

Regarding fatal vs. non-fatal errors in general. They both have pros and cons, and we need some sort of a (informal) policy which one to use in each case.

Reply:
The original policy was, at least for the type checker: If we cannot determine the correct type of the expression, it is a fatal error. This was relaxed with the push for type checker recovery, but as we see, it is very hard to get right.